CMake/IntelCompilers.cmake

```diff
@@ -1,79 +1,132 @@
-  message(FATAL_ERROR "Requires Intel 19 update 3 (19.0.0.20190206) or higher!")
+message(DEBUG "CMAKE_CXX_COMPILER_VERSION ${CMAKE_CXX_COMPILER_VERSION}")
+if(CMAKE_CXX_COMPILER_ID MATCHES "IntelLLVM")
+  if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS 2021.3)
+    message(FATAL_ERROR "Requires Intel oneAPI 2021.3 or higher!")
+  endif()
+elseif(INTEL_ONEAPI_COMPILER_FOUND)
+  # in this case, the version string reported based on Clang, not accurate enough. just skip check.
+else()
+  if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS 19.0.0.20190206)
+    message(FATAL_ERROR "Requires Intel 19 update 3 (19.0.0.20190206) or higher!")
+  endif()
 endif()
 
 # Enable OpenMP
 if(QMC_OMP)
   set(ENABLE_OPENMP 1)
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -qopenmp")
-  if(ENABLE_OFFLOAD)
-    set(OFFLOAD_TARGET
-        "host"
-        CACHE STRING "Offload target architecture")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -qopenmp -qopenmp-offload=${OFFLOAD_TARGET}")
-  else(ENABLE_OFFLOAD)
+  if(CMAKE_CXX_COMPILER_ID MATCHES "IntelLLVM" OR INTEL_ONEAPI_COMPILER_FOUND)
+    if(ENABLE_OFFLOAD)
+      set(OFFLOAD_TARGET
+          "spir64"
+          CACHE STRING "Offload target architecture")
+      set(OPENMP_OFFLOAD_COMPILE_OPTIONS "-fopenmp-targets=${OFFLOAD_TARGET}")
+    endif(ENABLE_OFFLOAD)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fiopenmp")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fiopenmp")
+  else()
+    if(ENABLE_OFFLOAD)
+      message(FATAL_ERROR "OpenMP offload requires using Intel oneAPI compilers.")
+    endif()
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -qopenmp")
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -qopenmp")
-  endif(ENABLE_OFFLOAD)
+  endif()
 endif(QMC_OMP)
 
-# Suppress compile warnings
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-deprecated")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-deprecated")
+if(CMAKE_CXX_COMPILER_ID MATCHES "IntelLLVM" OR INTEL_ONEAPI_COMPILER_FOUND)
+  # oneAPI compiler options
 
-# Set extra optimization specific flags
-set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -restrict -unroll -ip")
-set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -restrict -unroll -ip")
-set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -restrict -unroll -ip")
-set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -restrict -unroll -ip")
-set(CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO} -restrict -unroll -ip")
-set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} -restrict -unroll -ip")
+  # Set clang specific flags (which we always want)
+  add_compile_definitions(restrict=__restrict__)
 
-# Set prefetch flag
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -qopt-prefetch")
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fstrict-aliasing")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fstrict-aliasing -D__forceinline=inline")
 
-if(MIXED_PRECISION)
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -prec-sqrt")
+  # Set extra optimization specific flags
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fomit-frame-pointer")
```

Comment:
as mentioned in #3300 a debug build with -fomit-frame-pointer is no desirable.

Reply:
Changed. So only debug build gets `-fno-omit-frame-pointer` added. In other cases, stick to the compiler default.